src/mod/server.mod/servmsg.c

```diff
@@ -1132,6 +1141,115 @@ static int got311(char *from, char *msg)
   return 0;
 }
 
+static int gotauthenticate(char *from, char *msg)
+{
+  char src[256] = ""; // FIXME: size
+  char *s;
+  size_t slen;
+  unsigned char dst[256] = ""; // FIXME: size
+  size_t olen;
+  unsigned char dst2[256] = ""; // FIXME: size
+  unsigned int olen2;
+#ifdef HAVE_OPENSSL_SSL_H
+  FILE *fp;
+  EC_KEY *eckey;
+  EVP_PKEY *privateKey;
+#endif
+
+  putlog(LOG_SERV, "*", "SASL: got AUTHENTICATE %s", msg);
+  if (msg[0] == '+') {
+    s = src;
+    /* Don't use snprintf due to \0 inside */
+    if (!strcmp(sasl_mechanism, "PLAIN")) {
+      strcpy(s, sasl_username);
+      s += strlen(sasl_username) + 1;
+      strcpy(s, sasl_username);
+      s += strlen(sasl_username) + 1;
+      strcpy(s, sasl_password);
+      s += strlen(sasl_password);
+    }
+    else if (!strcmp(sasl_mechanism, "ECDSA-NIST256P-CHALLENGE")) {
+      strcpy(s, sasl_username);
+      s += strlen(sasl_username) + 1;
+      strcpy(s, sasl_username);
+      s += strlen(sasl_username);
+    }
+    else /* sasl_mechanism == EXTERNAL */
+      *s++ = '+';
+    slen = s - src;
+    mbedtls_base64_encode(dst, sizeof dst, &olen, (const unsigned char *) src, slen);
+    putlog(LOG_SERV, "*", "SASL: put AUTHENTICATE %s", dst);
+    dprintf(DP_MODE, "AUTHENTICATE %s\n", dst);
+  } else {
+    putlog(LOG_SERV, "*", "SASL: got AUTHENTICATE Challange");
+    mbedtls_base64_decode(dst, sizeof dst, &olen, (const unsigned char *) msg, strlen(msg));
+    fp = fopen(sasl_key, "r");
+    if (!fp) {
+      putlog(LOG_SERV, "*", "SASL: AUTHENTICATE: fopen(): %s\n", sasl_key);
+      return 1; /* FIXME: 1 or 0 ? */
+    }
+    privateKey = PEM_read_PrivateKey(fp, NULL, 0, NULL);
+    if (!privateKey) {
+      putlog(LOG_SERV, "*", "SASL: AUTHENTICATE: PEM_read_PrivateKey(): SSL error = %s\n",
+             ERR_error_string(ERR_get_error(), 0));
+      fclose(fp);
+      return 1; /* FIXME: 1 or 0 ? */
+    }
+    fclose(fp);
+    eckey = EVP_PKEY_get1_EC_KEY(privateKey);
+    if (!eckey) {
+      putlog(LOG_SERV, "*", "SASL: AUTHENTICATE: EVP_PKEY_get1_EC_KEY(): SSL error = %s\n",
+             ERR_error_string(ERR_get_error(), 0));
+      return 1;
+    }
+    if (ECDSA_sign(0, dst, olen, dst2, &olen2, eckey) == 0) {
+      printf("SASL: AUTHENTICATE: ECDSA_sign() SSL error = %s\n",
+             ERR_error_string(ERR_get_error(), 0));
+      return 1;
+    } 
+    mbedtls_base64_encode(dst, sizeof dst, &olen, dst2, olen2);
+    putlog(LOG_SERV, "*", "SASL: put AUTHENTICATE Response %s", dst);
+    dprintf(DP_MODE, "AUTHENTICATE %s\n", dst);
+
+  }
+  return 1; /* FIXME: 1 or 0 ? */
+}
+
+static int got900(char *from, char *msg)
+{
+  newsplit(&msg); /* nick */
+  newsplit(&msg); /* nick!ident@host */
+  newsplit(&msg); /* account */
+  fixcolon(msg);
+  putlog(LOG_SERV, "*", "SASL: %s", msg);
+  return 1;
+}
+
+static int got903(char *from, char *msg)
+{
+  newsplit(&msg); /* nick */
+  fixcolon(msg);
+  putlog(LOG_SERV, "*", "SASL: %s", msg);
+  dprintf(DP_MODE, "CAP END\n");
```

Comment:
So yes, the SASL spec says to hold off on END until the SASL exchange is completed, but the CAP spec says END should be sent when capability negotiation is done (which I interpret as agnostic from the success or failure of implementing a capability). I think these two statements are at odds with each other. On the bright side, an END sent twice is ignored by the server. We probably need to handle the ENDs in the cap side of the house, so that we can properly track when to send it based on the needs of all requests?

Reply:
so, nothing to do here and now. you already placed a comment into the code, for future development when more caps will be handled by eggdrop. for now were fine.